File: Lista de Exercicios (Pilhas e Filas)/Filas/4_ImpressoraSim.py
from collections import deque

class Impressora:
    
    def __init__(self):
        
        self.fila = deque()

    def adicionar_documento(self, nome, paginas):
        self.fila.append({"nome": nome, "paginas": paginas})
        print(f"Documento '{nome}' com {paginas} páginas adicionado à fila.")

    def imprimir_proximo(self):
        if not self.fila:
            print("Nenhum documento na fila para imprimir.")
            return None

        documento = self.fila.popleft()
        print(f"Imprimindo documento '{documento['nome']}' com {documento['paginas']} páginas.")
        return documento

    def mostrar_fila(self):
        if not self.fila:
            print("A fila de impressão está vazia.")
        else:
            print("Fila de impressão:")
            for i, doc in enumerate(self.fila, start=1):
                print(f"{i}. {doc['nome']} - {doc['paginas']} páginas")


impressora = Impressora()

impressora.adicionar_documento("Trabalho", 10)
impressora.adicionar_documento("Relatório", 5)
impressora.adicionar_documento("Apresentação", 15)

impressora.mostrar_fila()

impressora.imprimir_proximo()
impressora.imprimir_proximo()

impressora.mostrar_fila()

impressora.imprimir_proximo()

impressora.imprimir_proximo()

File: Lista de Exercicios - C/Filas/4_ImpressoraSim.c
#include <stdio.h>
#include <string.h>

#define MAX 10
#define TAM_NOME 50

typedef struct {
    char nome[TAM_NOME];
    int paginas;
} Documento;

typedef struct {
    Documento fila[MAX];
    int inicio, fim, tamanho;
} Impressora;

void inicializar(Impressora *imp) {
    imp->inicio = 0;
    imp->fim = -1;
    imp->tamanho = 0;
}

void adicionar_documento(Impressora *imp, const char *nome, int paginas) {
    if (imp->tamanho == MAX) {
        printf("Fila cheia!\n");
        return;
    }
    imp->fim = (imp->fim + 1) % MAX;
    strcpy(imp->fila[imp->fim].nome, nome);
    imp->fila[imp->fim].paginas = paginas;
    imp->tamanho++;
    printf("Documento '%s' com %d páginas adicionado à fila.\n", nome, paginas);
}

void imprimir_proximo(Impressora *imp) {
    if (imp->tamanho == 0) {
        printf("Nenhum documento na fila para imprimir.\n");
        return;
    }
    Documento *doc = &imp->fila[imp->inicio];
    printf("Imprimindo documento '%s' com %d páginas.\n", doc->nome, doc->paginas);
    imp->inicio = (imp->inicio + 1) % MAX;
    imp->tamanho--;
}

void mostrar_fila(Impressora *imp) {
    if (imp->tamanho == 0) {
        printf("A fila de impressão está vazia.\n");
        return;
    }
    printf("Fila de impressão:\n");
    int i, idx;
    for (i = 0, idx = imp->inicio; i < imp->tamanho; i++, idx = (idx + 1) % MAX) {
        printf("%d. %s - %d páginas\n", i+1, imp->fila[idx].nome, imp->fila[idx].paginas);
    }
}

int main() {
    Impressora impressora;
    inicializar(&impressora);
    adicionar_documento(&impressora, "Trabalho", 10);
    adicionar_documento(&impressora, "Relatório", 5);
    adicionar_documento(&impressora, "Apresentação", 15);
    mostrar_fila(&impressora);
    imprimir_proximo(&impressora);
    imprimir_proximo(&impressora);
    mostrar_fila(&impressora);
    imprimir_proximo(&impressora);
    imprimir_proximo(&impressora);
    return 0;
}